### PR TITLE
create built in way to visualize dataframe in Jupyter

### PR DESCRIPTION
**Pull request by rgbkrk**
Hi! 👋🏻

This introduces a basic way to display Polars DataFrames from a deno kernel in a Jupyter Notebook.

image

While I've [baked it into `display.js`](https://github.com/rgbkrk/display.js/blob/main/test-notebooks/polars-df.ipynb), which is on its way to be part of `Deno` itself (https://github.com/denoland/deno/pull/20819), I'd love to get a solid implementation inside of `nodejs-polars`. There are two mediatypes exported here via the [Rich Content Output Symbol, `Symbol.for("Jupyter.display")`](https://docs.deno.com/runtime/manual/tools/jupyter#rich-content-output)

* `text/html` - a standard static table we're all pretty used to
* `application/vnd.dataresource+json` - Row oriented JSON for frontends to implement scrollable grids for a table (and additional data visualization).

Here's an example of what's possible with the data resource media type:

image

Blog post featuring `nodejs-polars`: https://blog.jupyter.org/bringing-modern-javascript-to-the-jupyter-notebook-fc998095081e

Once this is in I'll update the blog post screenshots to use the implementation that relies on `nodejs-polars`.

## Background on the Data Resource Schema

Using the schema from https://specs.frictionlessdata.io/schemas/tabular-data-resource.json, we can emit a format that frontends can use for much richer tables.

As an example, in `pandas` you enable it with:

```python
pd.set_option("display.html.table_schema", True)
```

and then just type `df` in cells. Here's some more sample (python) code for looking at this format

```python
import pandas as pd

df = pd.read_csv(
    "https://raw.githubusercontent.com/allisonhorst/palmerpenguins/main/inst/extdata/penguins.csv"
)
# Enable { schema, data } output as JSON
pd.set_option("display.html.table_schema", True)

# Now to demonstrate what's inside
import json

ip = get_ipython()

# Enable { schema, data } output as JSON
pd.set_option("display.html.table_schema", True)
data, metadata = ip.display_formatter.format(df.head(5))

data_resource = data['application/vnd.dataresource+json']

print(json.dumps(data_resource, indent=2))
```

```json
{
  "schema": {
    "fields": [
      {
        "name": "index",
        "type": "integer"
      },
      {
        "name": "species",
        "type": "string"
      },
      {
        "name": "island",
        "type": "string"
      },
      {
        "name": "bill_length_mm",
        "type": "number"
      },
      {
        "name": "bill_depth_mm",
        "type": "number"
      },
      {
        "name": "flipper_length_mm",
        "type": "number"
      },
      {
        "name": "body_mass_g",
        "type": "number"
      },
      {
        "name": "sex",
        "type": "string"
      },
      {
        "name": "year",
        "type": "integer"
      }
    ],
    "primaryKey": [
      "index"
    ],
    "pandas_version": "1.4.0"
  },
  "data": [
    {
      "index": 0,
      "species": "Adelie",
      "island": "Torgersen",
      "bill_length_mm": 39.1,
      "bill_depth_mm": 18.7,
      "flipper_length_mm": 181.0,
      "body_mass_g": 3750.0,
      "sex": "male",
      "year": 2007
    },
    {
      "index": 1,
      "species": "Adelie",
      "island": "Torgersen",
      "bill_length_mm": 39.5,
      "bill_depth_mm": 17.4,
      "flipper_length_mm": 186.0,
      "body_mass_g": 3800.0,
      "sex": "female",
      "year": 2007
    },
    {
      "index": 2,
      "species": "Adelie",
      "island": "Torgersen",
      "bill_length_mm": 40.3,
      "bill_depth_mm": 18.0,
      "flipper_length_mm": 195.0,
      "body_mass_g": 3250.0,
      "sex": "female",
      "year": 2007
    },
    {
      "index": 3,
      "species": "Adelie",
      "island": "Torgersen",
      "bill_length_mm": null,
      "bill_depth_mm": null,
      "flipper_length_mm": null,
      "body_mass_g": null,
      "sex": null,
      "year": 2007
    },
    {
      "index": 4,
      "species": "Adelie",
      "island": "Torgersen",
      "bill_length_mm": 36.7,
      "bill_depth_mm": 19.3,
      "flipper_length_mm": 193.0,
      "body_mass_g": 3450.0,
      "sex": "female",
      "year": 2007
    }
  ]
}
```

### CODE REVIEW

**Comment by rgbkrk** [comment]
Note: I could not actually build this branch [due to an issue with local builds, as mentioned on Discord](https://discord.com/channels/908022250106667068/908980667927396382/1161464311730937977). 

Hey all! I'm trying to make a PR to nodejs-polars and haven't been able to figure out how to get the toolchain working well:

```
...
active toolchain
----------------

nightly-2023-07-27-aarch64-apple-darwin (overridden by '/Users/kai/code/src/github.com/pola-rs/nodejs-polars/rust-toolchain')
rustc 1.73.0-nightly (0d95f9132 2023-07-26)


nodejs-polars on  main [$!?] is 📦 v0.8.2 via ⬢ v18.12.0 via 🦀 v1.72.1 took 24s
➜ yarn build && yarn build:ts
...
   Compiling polars-lazy v0.32.0 (https://github.com/pola-rs/polars.git?rev=ec0c91f93fcd1ac355c667d6c3c3f30b257ea0a6#ec0c91f9)
error[E0554]: `#![feature]` may not be used on the stable release channel
  --> /Users/kai/.cargo/registry/src/index.crates.io-6f17d22bba15001f/argminmax-0.6.1/src/lib.rs:72:39
   |
72 | #![cfg_attr(feature = "nightly_simd", feature(stdsimd))]
   |                                       ^^^^^^^^^^^^^^^^

error[E0554]: `#![feature]` may not be used on the stable release channel
  --> /Users/kai/.cargo/registry/src/index.crates.io-6f17d22bba15001f/argminmax-0.6.1/src/lib.rs:73:39
   |
73 | #![cfg_attr(feature = "nightly_simd", feature(avx512_target_feature))]
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0554]: `#![feature]` may not be used on the stable release channel
  --> /Users/kai/.cargo/registry/src/index.crates.io-6f17d22bba15001f/argminmax-0.6.1/src/lib.rs:74:39
   |
74 | #![cfg_attr(feature = "nightly_simd", feature(arm_target_feature))]
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0554]: `#![feature]` may not be used on the stable release channel
  --> /Users/kai/.cargo/registry/src/index.crates.io-6f17d22bba15001f/argminmax-0.6.1/src/lib.rs:72:47
   |
72 | #![cfg_attr(feature = "nightly_simd", feature(stdsimd))]
   |                                               ^^^^^^^

   Compiling enum_dispatch v0.3.12
...
```

Does anyone else run into this or know how to get to a decent state for a local build on MacOS? I'm extremely new to Rust so I might be missing something super simple.

**Comment by Bidek56** [comment]
@rgbkrk Do you want to add some documentation around this feature? Thx

**Comment by rgbkrk** [comment]
Totally. Where do you want it? Other than the README I'm only seeing jsdocs.

**Comment by Bidek56** [comment]
Probably in the Readme below webpack, unless you find a better place. Thx

**Comment by rgbkrk** [comment]
README is a great start! I'll get that submitted soon.

In a separate PR I'd love to get some user guides together to help people get started with polars in JavaScript / Typescript.

**Comment by Bidek56** [comment]
LGTM, but I am just wondering 2 things:
1. Is there a need for JS/TS DF library in the Jupyter ecosystem since everyone is using Python?
2. If we are adding `Deno` to our docs, do we need to add `Deno` to our CI tests? We are currently using Bun and Node.

**Comment by rgbkrk** [comment]
> Is there a need for JS/TS DF library in the Jupyter ecosystem since everyone is using Python?

Isn't... that what this library is?

I use Python everyday. There are also a *lot* of JavaScript developers out there and you've all made something that is really solid interop between the javascript ecosystem and arrow underneath. I can use D3 directly with Polars DataFrames to visualize data. In this [example polars notebook](https://github.com/rgbkrk/denotebooks/blob/main/Polar%20DataFrames.ipynb), I'm able to pass columns directly to D3.

image

I'm ready to enable JavaScript developers to use polars.

**Comment by Bidek56** [comment]
I think this library is more for backend service use in Node/Bun then DS, but that's up for discussion. :)

File: polars/dataframe.ts [hunk 1]
```diff
@@ -1755,6 +1801,66 @@ export const _DataFrame = (_df: any): DataFrame => {
     set columns(names) {
       _df.columns = names;
     },
+    /**
+     * Return back text/html and application/vnd.dataresource+json representations of the DataFrame
+     * @returns Media bundle / mimetype keys for Jupyter frontends
+     */
+    [Symbol.for("Jupyter.display")]() {
```
Comment (universalmind303) [inline polars/dataframe.ts:1808]: We could probably reuse some of this outside of just jupyter. Would you be willing to refactor the html rendering into a `df.toHTML()` function?

File: polars/dataframe.ts [hunk 2]
```diff
@@ -1755,6 +1801,66 @@ export const _DataFrame = (_df: any): DataFrame => {
...
+      // Add table headers
+      htmlTable += "<thead><tr>";
+      fields.forEach((field) => {
+        htmlTable += `<th>${escapeHTML(field.name)}</th>`;
+      });
+      htmlTable += "</tr></thead>";
+
+      // Add table data
+      htmlTable += "<tbody>";
+      df.head(10)
```
Comment (universalmind303) [inline polars/dataframe.ts:1848]: why do we get the head of `50,` then later get the head of `10` here? `toRecords` is one of the most expensive operations, we probably shouldn't call it twice unless we really need to.

File: polars/dataframe.ts [hunk 3]
```diff
@@ -1755,6 +1801,66 @@ export const _DataFrame = (_df: any): DataFrame => {
...
+      const fields = this.getColumns().reduce(
+        (acc: DataResourceField[], curr) => {
+          acc.push({
+            name: curr.name,
+            type: mapPolarsTypeToJSONSchema(curr.dtype),
+          });
+          return acc;
+        },
+        [],
+      );
```
Comment (universalmind303) [inline polars/dataframe.ts:1824]: couldn't you just do a `map` here instead of a `reduce`? 

```js
const fields = this.getColumns().map(column => ({
            name: column.name,
            type: mapPolarsTypeToJSONSchema(column.dtype),
          })
```

**Review by universalmind303 — CHANGES_REQUESTED** [review]
great work @rgbkrk. Really excited to see more data science & analytics tools embracing JS. (_ I personally strongly dislike python syntax_).

Also excited to see that polars works on deno now. I tried using polars with deno a long time ago & it was unsupported.

Just a few small comments on the PR.

File: polars/dataframe.ts (hunk 1 above)
Comment (rgbkrk) [reply polars/dataframe.ts:1808]: Totally.

File: polars/dataframe.ts (hunk 3 above)
Comment (rgbkrk) [reply polars/dataframe.ts:1824]: So true, thank you.

File: polars/dataframe.ts (hunk 2 above)
Comment (rgbkrk) [reply polars/dataframe.ts:1848]: Oops. I should be using the already created records from above. For the HTML side I kept the table size smaller.

Should I rely on `POLARS_FMT_MAX_ROWS` for all of these?

File: polars/dataframe.ts [hunk 4]
```diff
@@ -1707,6 +1704,55 @@ function map(df: DataFrame, fn: (...args: any[]) => any[]) {
...
+
+const rawRe = new RegExp(`[${[...rawToEntity.keys()].join("")}]`, "g");
+
+/**
+ * Escapes text for safe interpolation into HTML text content and quoted attributes
+ */
+export function escapeHTML(str: string) {
+  return str.replaceAll(rawRe, (m) => {
+    const entity = rawToEntity.get(m);
+    return entity !== undefined ? entity : m;
```
Comment (Bidek56) [inline polars/dataframe.ts:1752]: I think we can shorten this to: `return entity ?? m`

File: polars/dataframe.ts [hunk 5]
```diff
@@ -1707,6 +1704,55 @@ function map(df: DataFrame, fn: (...args: any[]) => any[]) {
...
+    List: "array",
+    Struct: "object",
+  };
+
+  const dataType = colType.variant;
+  return typeMapping[dataType] || "string";
+}
+
+/////// HTML ESCAPING ///////
+const rawToEntityEntries = [
```
Comment (Bidek56) [inline polars/dataframe.ts:1734]: Any reason this is a global const but only used inside of `escapeHTML` func?
Should we move it inside of the function?

File: polars/dataframe.ts [hunk 6]
```diff
@@ -1707,6 +1704,55 @@ function map(df: DataFrame, fn: (...args: any[]) => any[]) {
...
+  ["&", "&amp;"],
+  ["<", "&lt;"],
+  [">", "&gt;"],
+  ['"', "&quot;"],
+  ["'", "&#39;"],
+] as const;
+
+const rawToEntity = new Map<string, string>(rawToEntityEntries);
+
+const rawRe = new RegExp(`[${[...rawToEntity.keys()].join("")}]`, "g");
```
Comment (Bidek56) [inline polars/dataframe.ts:1744]: Should we move this inside of `escapeHTML` func?

File: polars/dataframe.ts (hunk 2 above)
Comment (Bidek56) [reply polars/dataframe.ts:1848]: We should check for `POLARS_FMT_MAX_ROWS` and `POLARS_FMT_MAX_COLS` but have a large max (10K rows) so it does not blow up the browser.
I usually use `df.limit(...)` or `df.head(..)` so that should work.

File: polars/dataframe.ts (hunk 1 above)
Comment (Bidek56) [reply polars/dataframe.ts:1808]: How long before someone asks for a CSS option to this func? :)

Comment (rgbkrk) [reply polars/dataframe.ts:1808]: I hope at that point they realize how to use `toRecords` themselves.

File: polars/dataframe.ts (hunk 2 above)
Comment (rgbkrk) [reply polars/dataframe.ts:1848]: I now realize there's a bit of conflicting advice in here, depending on how I implement this.

If I break up `toHTML` and `toDataResource` to individual functions then they'll need to both call `toRecords`. If I do it all in one display function, it can create the records just once. This could be a separate function that accepts the records too.

I will of course admit that it would be better for this to be a column-oriented emission but these are the protocols we have today until I can push for more Arrow everywhere.

File: polars/dataframe.ts (hunk 4 above)
Comment (rgbkrk) [reply polars/dataframe.ts:1752]: Sweet. Originally I had this as a non-null assertion and then typescript complained about it with the current rules. This will work.

File: polars/dataframe.ts (hunk 5 above)
Comment (rgbkrk) [reply polars/dataframe.ts:1734]: Style wise I understand. However this will create it on every call, which is called for _every_ cell in the table output.

I think there's an inner dependency that relies on `escape-html` I could add in as an explicit dependency if that's better.

File: polars/dataframe.ts (hunk 2 above)
Comment (universalmind303) [reply polars/dataframe.ts:1862]: To elaborate, I think it'd be cool to make the `toHTML()` function behave similarly to `toRecords()` _(with no limit/head on it)_. Then use that inside of here. It should make this much more concise while also providing reusability via `toHTML()` 

```ts
[jupyterDisplay]() {
  const html = this.limit(50).toHTML(); 
  const fields = ...
  return {
    "application/vnd.dataresource+json": { data, schema: { fields } },
    "text/html": htmlTable,
  };
}

Comment (rgbkrk) [reply polars/dataframe.ts:1862]: Ok, cool will do.

**Review by universalmind303 — COMMENTED** [review]
Can we add a test for this? otherwise looks like it's good to go!

**Comment by Bidek56** [comment]
@universalmind303 Do we need a CI test using `Deno` now that we are using it?

**Comment by rgbkrk** [comment]
Ok, the test is in and this is working really nicely.

---

🦕 I'm not sure how you're going to want to test it with Deno but it seems like a good follow up PR.

The way I'm exercising this right now for local hacking is with a `mod.ts` (for Deno) that contains:

```typescript
import { createRequire } from "node:module"

const require = createRequire(import.meta.url)

export const pl = require('./bin/index.js');

export default pl;
```

When this is shipped we'll be able to just use `import pl from npm:nodejs-polars`. I only have this `mod.ts` to make it easy to work with the local node package.

CC @sigmaSd @bartlomieju

**Comment by universalmind303** [comment]
Thanks again @rgbkrk. I'm very excited to see more JS + notebook integrations maturing.

**Comment by rgbkrk** [comment]
Thank you for the review and the stewardship of polars for JavaScript! I'm excited to see what people can do with this.
